sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1831,8 +1838,8 @@ object PushPredicateThroughNonJoin extends Rule[LogicalPlan] with PredicateHelpe
       }
 
       if (pushDown.nonEmpty) {
-        val pushDownPredicate = pushDown.reduce(And)
```

Comment:
so the reason we don't simply rewrite the AND combined predicate is that we want to keep the predicate splittable later on?

Reply:
Yes

Reply:
But then there is no common expression sharing between sub-predicates split by AND.

Shall we simply support splitting the predicate expression that is wrapped by `With`?

Reply:
e.g. for `a * a > 0 AND a + a > 0` and `a` is a large expression, we first push down `With(large_expr, a * a > 0 AND a + a > 0)`, when we need to split it again later, we turn it into `With(large_expr, a * a > 0) AND With(large_expr, a + a > 0)`

Reply:
OK I think this will go very complicated. We can't completely avoid duplicating large expressions because at some point, we can push down part of the predicates, so the common expression will appear in multiple plan nodes and we don't support cross-plan-nodes expression sharing yet.

Reply:
`select add,count(1) as ct from (select a+a as add,b from t) where add + add > 10 and add >ct`, 
the current implementation can push down add+add > 10 to aggregate.

Reply:
More thoughts:
- I think it's common to reference the same column in multiple sub-predicates, e.g. `a > 0 AND a < 10`. The current approach doesn't help with this common case.
- We should improve the predicate split-and-combine process, to only accept duplicating large expressions when we have to, i.e. some sub-predicates need to be pushed down and some need to stay, and they share the same common expressions.

Reply:
In the future, we should also support cross-plan-nodes expression sharing, by adding a Project to evaluate the common expression and propagate the result column all the way up.

Reply:
> We should improve the predicate split-and-combine process, to only accept duplicating large expressions when we have to, i.e. some sub-predicates need to be pushed down and some need to stay, and they share the same common expressions.

In this follow-up, we can merge the with expression before rewriteWith. This will not affect the push down.

Reply:
> In the future, we should also support cross-plan-nodes expression sharing, by adding a Project to evaluate the common expression and propagate the result attribute all the way up.


Yes, this is the ideal state. I planned to do this at first, so I made a lot of changes to with and encountered many problems.

Reply:
> In this follow-up, we can merge the with expression before rewriteWith. This will not affect the push down.

It's better to not duplicate expressions in the first place. Creating a huge plan tree is harmful to the optimizer/planner.

Reply:
What I mean is that when cross-plan-nodes expression sharing is not supported, after the pushdown is completed and before rewriteWith, the with in the same filter can be merged.

Reply:
OK, It seems to be more troublesome, a + a> 5 and a < 10, a + a> 5 generates with, a < 10 is expanded.